docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeAccount`
+* `InitializeMultisig` -- TODO deprecate this and use a more general multisig?
+* `Revoke`
+* `SetAuthority`
+* `CloseAccount`
+* `FreezeAccount`
+* `ThawAccount`
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeAccount2`
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+
+New instructions required:
+
+* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+The `amount` field in an SPL token holding should contain a balance in order
+to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.
+
+### Token Program Registry
+
+We need the [token-list](https://github.com/solana-labs/token-list) to include
+vetted SPL token-conforming programs.
+
+TODO consider adding Rust and C versions of the registry for on-chain programs and RPC
+
+TODO consider detailing the process of including a new program
+
+### Runtime
+
+#### Ephemeral Accounts
```

Comment:
maybe like this?: https://github.com/solana-labs/solana-program-library/tree/master/shared-memory

Reply:
Yes, the concept is based on the shared memory program.  I'm worried that the shared memory program adds a lot of friction for on-chain programs though -- it means that every instruction that uses tokens must *also* pass in a shared memory account and the shared memory program.  To make things easier, let's add it directly into the runtime.  What do you think?

Reply:
yeah, you have a point. and indeed you have a good point for the friction of future upgradability of underlying programs.

Maybe, that's another selling point of my idea of dynamic AccountInfo construction via preflight? :)

Reply:
That means even if we add some nice run-time-sanctioned special program, existing programs might not be able to use it for the possibility of breaking the AccountInfo list compatibility.